### PR TITLE
nixos/libvirtd: support out-of-tree vhost-user drivers

### LINKED ISSUE #187078: nixos/libvirt: Unable to find a satisfying virtiofsd
### Describe the bug
libvirt fails to start KVM domains when a virtiofs filesystem is present.

### Steps To Reproduce
1. Enable libvirtd
2. Run `virt-install -n scratch --connect qemu:///system --os-variant nixos-unstable --memory 1536 --memorybacking=access.mode=shared --disk none --import --filesystem /nix/store/,nix-store,driver.type=virtiofs`
3. `ERROR    operation failed: Unable to find a satisfying virtiofsd`

### Expected behavior
It should "just work" for good UX

### Additional context
This probably requires patching libvirt. I don't understand the codebase enough to do so.

I tried upgrading [libvirt to 8.5.0,](https://github.com/lourkeur/distro/commit/e97a3dcd9b2b06703e1da4391460a2d181de8210) it doesn't seem to change anything.

There is a workaround:

> Maintaining `<binary path='/run/current-system/sw/bin/virtiofsd' xattr='on'>` through all XML's seems ugly

_Originally posted by @kvtb in https://github.com/NixOS/nixpkgs/issues/113172#issuecomment-1114183820_

### Notify maintainers

@fpletz @globin @lovesegfault @Kranzes 

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.59, NixOS, 22.05 (Quokka), 22.05.20220815.3d47bba`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - nixpkgs: `/etc/nix/channels/nixos`
```

### PR DESCRIPTION
**Pull request by alois31**
## Description of changes

Add support for out-of-tree vhost-user drivers to the libvirtd module. In addition, include the vhost-user configuration file in the virtiofsd package so that this option can be used for virtiofs.

Fixes: https://github.com/NixOS/nixpkgs/issues/187078

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/servers/misc/virtiofsd/default.nix [hunk 1]
```diff
@@ -21,6 +21,15 @@ rustPlatform.buildRustPackage rec {
 
   buildInputs = [ libcap_ng libseccomp ];
 
+  postBuild = ''
+    sed -i "s|/usr/libexec|$out/bin|g" 50-virtiofsd.json
+  '';
+
+  postInstall = ''
+    install -dm755 "$out/share/qemu/vhost-user"
+    install -m644 50-virtiofsd.json "$out/share/qemu/vhost-user/"
```
Comment (alyssais) [inline pkgs/servers/misc/virtiofsd/default.nix:30]: You can do this in one line with `install -D`.

File: pkgs/servers/misc/virtiofsd/default.nix [hunk 2]
```diff
@@ -21,6 +21,15 @@ rustPlatform.buildRustPackage rec {
 
   buildInputs = [ libcap_ng libseccomp ];
 
+  postBuild = ''
```
Comment (alyssais) [inline pkgs/servers/misc/virtiofsd/default.nix:24]: doesn't really matter, postConfigure would make more sense, as this is something that would be handled in the configure phase of a build system.

**Comment by alyssais** [comment]
virtiofsd changes lgtm

File: nixos/modules/virtualisation/libvirtd.nix [hunk 3]
```diff
@@ -502,6 +511,15 @@ in
...
     systemd.sockets.libvirtd.wantedBy = [ "sockets.target" ];
 
+    systemd.tmpfiles.rules = let
+      vhostUserCollection = pkgs.buildEnv {
+        name = "vhost-user";
+        paths = cfg.qemu.vhostUserPackages;
+        pathsToLink = [ "/share/qemu/vhost-user" ];
+      };
+    in optional (cfg.qemu.vhostUserPackages != [ ])
+    "L+ /var/lib/qemu/vhost-user - - - - ${vhostUserCollection}/share/qemu/vhost-user";
```
Comment (astro) [inline nixos/modules/virtualisation/libvirtd.nix:521]:
```suggestion
      "L+ /var/lib/qemu/vhost-user - - - - ${vhostUserCollection}/share/qemu/vhost-user";
```

**Review by astro — APPROVED** [review]

**Review by astro — APPROVED** [review]

File: nixos/modules/virtualisation/libvirtd.nix [hunk 4]
```diff
@@ -502,6 +511,15 @@ in
...
     systemd.sockets.libvirtd.wantedBy = [ "sockets.target" ];
 
+    systemd.tmpfiles.rules = let
+      vhostUserCollection = pkgs.buildEnv {
+        name = "vhost-user";
+        paths = cfg.qemu.vhostUserPackages;
+        pathsToLink = [ "/share/qemu/vhost-user" ];
+      };
+    in optional (cfg.qemu.vhostUserPackages != [ ])
+      "L+ /var/lib/qemu/vhost-user - - - - ${vhostUserCollection}/share/qemu/vhost-user";
```
Comment (alyssais) [inline nixos/modules/virtualisation/libvirtd.nix:521]: Any reason not to just always include this?

**Comment by alyssais** [comment]
@OfBorg test libvirtd

**Pull request merged**